### PR TITLE
Update Allocate method to reuse lease if present

### LINKED ISSUE #329: cni v0.8.x results in no networking for podman
CNI-0.8 has become available in several distributions and appears to be resulting in podman unable to configure networking.  I'm logging this issue to get the discussion going and perhaps we can collaboratively debug first this this week.

One observation, on a F30 VM with the latest podman in koji, i hand built the cni cplugins and that worked nicely. i wonder if our build of cni did something funny ?

### PR DESCRIPTION
**Pull request by EmilyShepherd**
Previously, the Allocate method of the daemon always created a new Lease object. However, as both the CNI ADD and CHECK commands call Allocate, and CHECK can be called multiple times, this resulted in multiple Lease objects being created per pod.

Each of these leases was long lived with its own maintain() loop - however the daemon only kept track of the most recent one, meaning any old lease objects remained running forever (and held open their NetNS files). After a long enough period, this resulted in the system crashing out with "too many files" or a similar error limits-related error.

This commit updates the behaviour of Allocate() to first check if a Lease already exists for the given clientID. If none is found, one is created as before. If a Lease is found, a new Check() mechanism is called, which simply wakes up the maintain() loop to cause it to check the status of the lease.

This may fix [#329](https://github.com/containernetworking/cni/issues/329).

Signed-off-by: Emily Shepherd <emily@redcoat.dev>

### CODE REVIEW

**Comment by EmilyShepherd** [comment]
NB: The test failure on this PR appears to be unrelated to these changes.

Please see https://github.com/containernetworking/plugins/pull/793 for a potential fix.

**Comment by EmilyShepherd** [comment]
Bump @dcbw 🙏

**Comment by mccv1r0** [comment]
Looks good.  I tested the PR via cnitool using bridge plugin and isc-dhcpd-server

```
2023/01/04 14:42:21 cnitool-1d56652e125f8a333e30/bridge-dhcp0-chain/ens0: lease acquired, expiration is 2023-01-04 14:45:41.855778598 -0500 EST m=+1489.727401399
2023/01/04 14:42:26 cnitool-1d56652e125f8a333e30/bridge-dhcp0-chain/ens0: Checking lease
2023/01/04 14:42:29 cnitool-1d56652e125f8a333e30/bridge-dhcp0-chain/ens0: Checking lease
2023/01/04 14:44:01 cnitool-1d56652e125f8a333e30/bridge-dhcp0-chain/ens0: renewing lease
2023/01/04 14:44:01 cnitool-1d56652e125f8a333e30/bridge-dhcp0-chain/ens0: lease renewed, expiration is 2023-01-04 14:47:21.883450554 -0500 EST m=+1589.755073355
2023/01/04 14:45:41 cnitool-1d56652e125f8a333e30/bridge-dhcp0-chain/ens0: renewing lease
2023/01/04 14:45:41 cnitool-1d56652e125f8a333e30/bridge-dhcp0-chain/ens0: lease renewed, expiration is 2023-01-04 14:49:01.990540685 -0500 EST m=+1689.862163459
2023/01/04 14:45:42 cnitool-1d56652e125f8a333e30/bridge-dhcp0-chain/ens0: Checking lease
2023/01/04 14:45:54 cnitool-1d56652e125f8a333e30/bridge-dhcp0-chain/ens0: Checking lease

```

**Review by mccv1r0 — APPROVED** [review]

File: plugins/ipam/dhcp/lease.go [hunk 1]
```diff
@@ -67,6 +67,7 @@ type DHCPLease struct {
 	broadcast     bool
 	stopping      uint32
 	stop          chan struct{}
+	check         chan bool
```
Comment (squeed) [inline plugins/ipam/dhcp/lease.go:70]: nit: it's convention to use the type `chan struct{}` to indicate that a channel carries no information.

**Comment by squeed** [comment]
Looks great! One minor nit, then we can get this merged.

**Review by squeed — APPROVED** [review]

**Comment by squeed** [comment]
Thanks!

**Pull request merged**